From 122f90eb0bdb46cf95cf94b11a9278262bfa4252 Mon Sep 17 00:00:00 2001
From: David Kershner <dkershner@habana.ai>
Date: Mon, 02 Jun 2025 16:30:32 +0300
Subject: [PATCH] [SW-226950] oobi: add sensor op state data

Need to persist sensor operational state that is set by PLDM commands
SetNumericSensorEnable and SetStateSensorEnables.  The oobi data
contains sensors that have been set to either DISABLED or UNAVAILABLE,
ENABLED sensors are not stored in the list, since all sensors by
default are enabled.

Note: We have a storage limit of 300 numeric sensors and 64 state
sensors due to space issues. If more than that is need to be set to
disabled or unavailble, the state will not be persisted across
resets.

Change-Id: Ic3b0dd08e7dc083e7ef1f8baf6a4d596289a3f46
Signed-off-by: David Kershner <dkershner@habana.ai>
---

diff --git a/oobi.c b/oobi.c
index a2c2a93..7963ff9 100644
--- a/oobi.c
+++ b/oobi.c
@@ -44,7 +44,7 @@
 #endif
 
 #define OOBI_MAGIC		0x4F4F4249 /* "OOBI" */
-#define OOBI_VERSION		3
+#define OOBI_VERSION		4
 #define OOBI_MIN_VERSION	1
 
 /* Out-of-Band Persistent Information (oobi)
@@ -122,6 +122,8 @@
 		struct hl_oobi_pldm_effecters pldm_effecter_data;
 		struct hl_oobi_pldm_sensors pldm_sensor_data;
 	} v3;
+
+	/* v4 - Additional field was added to bottom of pldm_sensor_data */
 };
 
 BUILD_ASSERT(sizeof(struct oobi) <= DT_REG_SIZE(DT_NODELABEL(oobi)),
diff --git a/oobi.h b/oobi.h
index ee572e9..b07ebfe 100644
--- a/oobi.h
+++ b/oobi.h
@@ -35,6 +35,8 @@
 #define OOBI_PLDM_MAX_NUM_NUMERIC_EFFECTERS	8
 #define OOBI_PLDM_MAX_NUM_STATE_EFFECTERS	16
 #define OOBI_PLDM_MAX_NUM_THRESHOLDS		64
+#define OOBI_PLDM_MAX_NUM_NUMERIC_SENSORS	300
+#define OOBI_PLDM_MAX_NUM_STATE_SENSORS		64
 
 struct hl_oobi_pldm_fu_component_state {
 	uint8_t state;
@@ -99,6 +101,16 @@
 	struct pldm_pmc_threshold_data thresholds;
 } __packed;
 
+struct hl_oobi_pldm_numeric_sensor_op_state {
+	uint16_t sensor_op_state:1;
+	uint16_t sensor_id:15;
+} __packed;
+
+struct hl_oobi_pldm_state_sensor_op_state {
+	uint16_t sensor_id;
+	uint16_t comp_op_state; /* 2 bits per comp_op_sensor */
+} __packed;
+
 /* Unordered collection effecter states, an effecter_id of 0 indicates
  * entry is empty. Does not indicate end of list.
  */
@@ -112,6 +124,10 @@
 	 * entry is empty. Does not indicate end of list.
 	 */
 	struct hl_oobi_pldm_sensor_threshold sensor_thresholds[OOBI_PLDM_MAX_NUM_THRESHOLDS];
+	/* v4 fields */
+	struct hl_oobi_pldm_numeric_sensor_op_state
+		numeric_op_state[OOBI_PLDM_MAX_NUM_NUMERIC_SENSORS];
+	struct hl_oobi_pldm_state_sensor_op_state state_op_state[OOBI_PLDM_MAX_NUM_STATE_SENSORS];
 } __packed;
 
 void hl_oobi_pldm_fu_restore_poc_data(struct hl_oobi_pldm_fu_poc_data *data);
